### ISSUE TITLE
clangd gets confused by `-U...` in command line

### ISSUE THREAD

**Issue by jpalus** [llvm/llvm-project#97759]
For input file `test.c`:
```c
#include <stdio.h>

#ifndef MSG
#define MSG "hello world"
#endif

void hello() {
  printf(MSG);
}
```
and `compile_commands.json`:
```json
[
  {      
    "directory": "/tmp",      
    "command": "clang -UMSG -c -o test.o test.c'",            
    "file": "test.c",            
    "output": "test.o"            
  }      
]
```
clangd reports false positive:
```console
$ clangd --check=test.c
...
I[21:05:56.452] Indexing headers...
I[21:05:56.472] Building AST...
E[21:05:56.491] [undeclared_var_use] Line 8: use of undeclared identifier 'MSG'
I[21:05:56.491] Indexing AST...
I[21:05:56.492] Building inlay hints
I[21:05:56.492] Building semantic highlighting
I[21:05:56.492] Testing features at each token (may be slow in large files)
I[21:05:56.493] Found definition heuristically using nearby identifier MSG
I[21:05:56.493] Found definition heuristically using nearby identifier MSG
I[21:05:56.501] All checks completed, 1 errors
```

**Comment by HighCommander4** [comment]
I believe this has the same underlying cause as https://github.com/clangd/clangd/issues/1396: "clang first loads the preamble and then applies the builtin buffer".

Here, the `#define MSG` becomes part of the preamble, and the `-UMSG` becomes part of the builtin buffer.

A workaround is to add a non-preprocessor declaration between the includes and the `#define MSG` to force the preamble to end before the `#define MSG`:

```c
#include <stdio.h>

struct EndPreamble {}; // dummy declaration to force the end of the preamble

#ifndef MSG
#define MSG "hello world"
#endif

void hello() {
  printf(MSG);
}
```

**Comment by HighCommander4** [comment]
Duplicate of https://github.com/clangd/clangd/issues/1396
